### PR TITLE
feat(2767): Handle stage logic in build workflow [5]

### PR DESCRIPTION
**Pull request by tkyi**
## Context

Properties might want to fetch stageBuilds for a stage.

## Objective

This PR adds an endpoint to get stageBuilds.

## References

Related to https://github.com/screwdriver-cd/screwdriver/issues/2767

## License

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: plugins/builds/index.js [hunk 1]
```diff
@@ -785,18 +820,86 @@ const buildsPlugin = {
...
+                    const workflowStageBuilds = [];
+
+                    current.flattenedWorkflowGraph.nodes.forEach(n => {
+                        if (n.stageName === currentNode.stageName) {
+                            workflowStageBuilds.push(n.name);
+                        }
+                    });
+
+                    // get all builds in current stage
+                    const buildParentBuilds = current.build.parentBuilds;
```
Comment (sagar1312) [inline plugins/builds/index.js:871]: Relying on `parentBuilds` may not work for all scenarios.
For example: Below stage has 6 jobs on two separate paths running in parallel
image
Parent builds for `prod-blue-certify` will only include `prod-blue-deploy` and `prod-blue-test`.
So the `teardown` will not be triggered when the all the jobs in the stage are successfully completed.

File: plugins/builds/index.js [hunk 2]
```diff
@@ -785,18 +820,86 @@ const buildsPlugin = {
...
+                } else {
+                    //  else get all stage builds in workflow, compare number of stageBuilds to stage builds with status
+                    // Get job names from stage workflow
+                    const workflowStageBuilds = [];
+
+                    current.flattenedWorkflowGraph.nodes.forEach(n => {
+                        if (n.stageName === currentNode.stageName) {
+                            workflowStageBuilds.push(n.name);
+                        }
+                    });
```
Comment (sagar1312) [inline plugins/builds/index.js:868]:
```suggestion
                    const workflowStageBuilds = currentStageBuild.workflowGraph.nodes.map(n => return n.name);
```

File: plugins/builds/index.js (hunk 1 above)
Comment (sagar1312) [reply plugins/builds/index.js:871]: Another scenario could be a stage with only one job.

File: plugins/builds/index.js [hunk 3]
```diff
@@ -785,18 +820,86 @@ const buildsPlugin = {
...
+                                const buildId = buildParentBuilds[pipeline.id].jobs[b];
+                                const buildForStage = eventBuilds.filter(eb => eb.id === buildId);
+
+                                stageDict[b] = buildForStage.length > 0 ? buildForStage[0] : undefined;
+                            }
+                        });
+                    }
+
+                    // Make sure number of builds is same as expected number of stage jobs
+                    if (Object.values(stageDict).every(b => b !== undefined)) {
```
Comment (sagar1312) [inline plugins/builds/index.js:887]: `teardown` may not get triggered in some scenarios.
For Ex: Below stage has 3 jobs
image
When `prod-blue-test` fails, `teardown` should get triggered. But this logic expect build for all the jobs within a stage. Since `prod-blue-certify` does not have a build, `teardown` will not be triggered.

File: plugins/builds/index.js [hunk 4]
```diff
@@ -785,18 +820,86 @@ const buildsPlugin = {
...
+                        // Check build statuses
+                        const stageIsDone = builds.every(b => isNotActiveBuild(b.status, b.endTime));
+                        const shouldRunStageTeardown = builds.some(b => hasTerminalBuild(b.status, b.endTime));
+
+                        // if any have terminal status or all no longer active, destroy created stage builds, set next job to teardown
+                        if (shouldRunStageTeardown || stageIsDone) {
+                            builds.forEach(async b => deleteBuild(b, buildFactory));
+                            nextJobsTrigger = [`stage@${currentNode.stageName}:teardown`];
+                        }
+                        // TODO: set final stage status
```
Comment (sagar1312) [inline plugins/builds/index.js:899]: Note to myself. To review this once implemented.

**Review by sagar1312 — CHANGES_REQUESTED** [review]

File: plugins/stages/stageBuilds/list.js [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+const boom = require('@hapi/boom');
+const joi = require('joi');
+const schema = require('screwdriver-data-schema');
+const listSchema = joi.array().items(schema.models.stageBuild.get).label('List of stage builds for a stage');
+const idSchema = schema.models.stage.base.extract('id');
+const eventIdSchema = schema.models.stageBuild.base.extract('eventId');
+
+module.exports = () => ({
+    method: 'GET',
+    path: '/stages/{id}/stageBuilds',
```
[the remark this answers is not in the document]
Comment (tkyi) [reply plugins/stages/stageBuilds/list.js:12]: added

File: plugins/builds/index.js [hunk 6]
```diff
@@ -765,7 +920,40 @@ const buildsPlugin = {
...
+                                    pipelineId: current.pipeline.id
+                                });
+
+                                if (nextStage) {
+                                    const nextStageBuild = await stageBuildFactory.get({
+                                        stageId: nextStage.id,
+                                        eventId: current.event.id
+                                    });
+
+                                    await nextStageBuild.remove();
```
Comment (sagar1312) [inline plugins/builds/index.js:953]: This should be abstracted into `build` model
https://github.com/screwdriver-cd/models/blob/master/lib/build.js#L652
```suggestion
                            const nextJobIsStageSetup = STAGE_SETUP_PATTERN.test(nextJobName);

                            // Delete next stageBuild if next job is stage setup
                            if (nextJobIsStageSetup) {
                                const [, nextStageName] = nextJobName.match(STAGE_SETUP_PATTERN);
                                const nextStage = await stageFactory.get({
                                    name: nextStageName,
                                    pipelineId: current.pipeline.id
                                });

                                if (nextStage) {
                                    const nextStageBuild = await stageBuildFactory.get({
                                        stageId: nextStage.id,
                                        eventId: current.event.id
                                    });

                                    await nextStageBuild.remove();
```

File: plugins/builds/update.js [hunk 7]
```diff
@@ -245,18 +268,58 @@ module.exports = () => ({
...
+                stageBuildHasFailure = ['FAILURE', 'ABORTED', 'UNSTABLE'].includes(stageBuild.status);
+            }
 
+            // Guard against triggering non-successful or unstable builds
+            // Don't further trigger pipeline if intend to skip further jobs
             if (newBuild.status !== 'SUCCESS' || skipFurther) {
                 // Check for failed jobs and remove any child jobs in created state
                 if (newBuild.status === 'FAILURE') {
-                    await removeJoinBuilds({ pipeline, job, build: newBuild }, request.server.app);
+                    await removeJoinBuilds(
```
Comment (sagar1312) [inline plugins/builds/update.js:303]: `removeJoinBuild()` needed to be invoked when the current build is `teardown` and has successfully completed, but the `stageBuild` status is a negative terminal state

File: plugins/builds/index.js [hunk 8]
```diff
@@ -870,19 +1058,77 @@ const buildsPlugin = {
...
+                            // If next build is stage setup, create stageBuild and set to running
+                            if (nextBuildStage && nextJobName === stageSetupName) {
+                                await stageBuildFactory.create({
+                                    stageId: nextBuildStage.id,
+                                    eventId: current.event.id,
+                                    status: 'CREATED'
+                                });
+                            }
+                        }
+
```
Comment (sagar1312) [inline plugins/builds/index.js:1086]: We should abstract this into `buildFactory`. This would ensure, `stageBuild` will always gets created when `setup` build is created irrespective of from where the API is called.

File: plugins/builds/index.js [hunk 9]
```diff
@@ -870,19 +1058,77 @@ const buildsPlugin = {
...
+                                await existNextBuild.remove();
+
+                                return null;
+                            }
+
+                            return existNextBuild;
+                        }
+                    }
+
+                    // Current build is not part of stage
```
Comment (sagar1312) [inline plugins/builds/index.js:1122]: With the latest changes in `index.js` update build endpoint, `triggerNextJobs()` will not be invoked when `teardown` has successfully completed, but `stageBuild` has failure. Hence, this block of code will not clean up the downstream builds of `teardown`.

Instead, invoke `removeJoinBuilds()` in `index.js` update build endpoint to cleanup downstream builds of `teardown` when `stageBuild` has failure.
```suggestion
                    if (!['CREATED', null, undefined].includes(existNextBuild.status)) {
                        return existNextBuild;
                    }
                    
```

File: plugins/builds/index.js [hunk 10]
```diff
@@ -515,6 +518,83 @@ async function getParentBuildStatus({ newBuild, joinListNames, pipelineId, build
...
+            start: false
+        });
+
+        return stageTeardownBuild;
+    }
+
+    // Create stageBuild if newBuild is stage setup
+    if (jobName === stageSetupName) {
+        await stageBuildFactory.create({ stageId: stage.id, eventId: event.id, status: 'CREATED' });
+    }
```
Comment (sagar1312) [inline plugins/builds/index.js:593]: This is a common step for both `AND` and `OR`. As suggested, if we abstract the logic to create `stageBuild` when `setup` build is created into `buildFactory`, we can remove this duplicate code.

File: plugins/builds/index.js [hunk 11]
```diff
@@ -515,6 +518,83 @@ async function getParentBuildStatus({ newBuild, joinListNames, pipelineId, build
...
+            jobName: stageTeardownName,
+            username,
+            scmContext,
+            event, // this is the parentBuild for the next build
+            baseBranch: event.baseBranch || null,
+            start: false
+        });
+
+        return stageTeardownBuild;
+    }
```
Comment (sagar1312) [inline plugins/builds/index.js:588]: If a stage job fails, `teardown` build would be created via `removeJoinBuilds` logic.
Why do we need to create `teardown` build here?

**Review by sagar1312 — CHANGES_REQUESTED** [review]

File: plugins/builds/index.js [hunk 12]
```diff
@@ -515,6 +518,83 @@ async function getParentBuildStatus({ newBuild, joinListNames, pipelineId, build
...
+                params: { eventId: event.id, jobId: stageTeardownJob[0].id }
+            });
+
+            if (existingStageTeardownBuild && existingStageTeardownBuild[0]) {
+                return existingStageTeardownBuild[0];
+            }
+        }
+
+        // Doesn't exist, create stage teardown job and return as next job
+        const stageTeardownBuild = await createInternalBuild({
```
Comment (sagar1312) [inline plugins/builds/index.js:575]: How would `teardown` build created here would get started/executed?

File: plugins/builds/index.js (hunk 8 above)
Comment (tkyi) [reply plugins/builds/index.js:1086]: Opened a PR: https://github.com/screwdriver-cd/models/pull/600

File: plugins/builds/index.js (hunk 10 above)
Comment (tkyi) [reply plugins/builds/index.js:593]: Opened a PR: https://github.com/screwdriver-cd/models/pull/600

File: plugins/builds/index.js [hunk 13]
```diff
@@ -709,6 +764,67 @@ function getParentBuildIds({ currentBuildId, parentBuilds, joinListNames, pipeli
...
+
+    if (nextJobName === stageTeardownName) {
+        return buildDeletePromises;
+    }
+    if (buildConfig.eventId) {
+        // Remove next build
+        buildDeletePromises.push(deleteBuild(buildConfig, buildFactory));
+
+        // Check if stage teardown build already exists
+        const stageTeardownJob = await jobFactory.get({
```
Comment (sagar1312) [inline plugins/builds/index.js:798]: We should separate the concerns. Let us not interleave the responsibility of creating `teardown` build along side cleaning up of join builds.

Can we move the logic to create `teardown` build to separate function `checkAndCreateStageTeardownBuild()` and invoke this when a stage job fails?

File: plugins/builds/index.js [hunk 14]
```diff
@@ -873,16 +1020,44 @@ const buildsPlugin = {
                         return createInternalBuild(internalBuildConfig);
                     }
 
-                    if (!['CREATED', null, undefined].includes(existNextBuild.status)) {
+                    if (!['CREATED', null, undefined].includes(existNextBuild.status) && !current.stage) {
```
Comment (sagar1312) [inline plugins/builds/index.js:1023]: As suggested here
https://github.com/screwdriver-cd/screwdriver/pull/2870/files/9f679937feaa4e278d80b3b1ef245984c9b03993#r1448111687, we need to revert this change.

```suggestion
                    if (!['CREATED', null, undefined].includes(existNextBuild.status)) {
```

File: plugins/builds/index.js [hunk 15]
```diff
@@ -873,16 +1020,44 @@ const buildsPlugin = {
...
+                                await existNextBuild.remove();
+
+                                return null;
+                            }
+
+                            return existNextBuild;
+                        }
+                    }
+
+                    // Current build is not part of stage
```
Comment (sagar1312) [inline plugins/builds/index.js:1053]: As suggested here https://github.com/screwdriver-cd/screwdriver/pull/2870/files/9f679937feaa4e278d80b3b1ef245984c9b03993#r1448111687, this block of code is no longer needed.

```suggestion
```

File: plugins/builds/update.js [hunk 16]
```diff
@@ -245,16 +270,99 @@ module.exports = () => ({
...
+            // (if stage teardown build exists, and stageBuild.status is negative,
+            // and there are no active stage builds, and teardown build is not started)
+            if (stage) {
+                const stageTeardownName = getFullStageJobName({ stageName: stage.name, jobName: 'teardown' });
+                const stageTeardownJob = await jobFactory.get({
+                    params: { pipelineId: pipeline.id, name: stageTeardownName }
+                });
+                let stageTeardownBuild;
+
+                if (stageTeardownJob) {
```
Comment (sagar1312) [inline plugins/builds/update.js:333]: This check is not needed. A stage must always have `teardown` job
```suggestion
               
```

File: plugins/builds/update.js [hunk 17]
```diff
@@ -245,16 +270,99 @@ module.exports = () => ({
...
+                await triggerNextJobs(
+                    { pipeline, job, build: newBuild, username, scmContext, event: newEvent, stage },
+                    request.server.app
+                );
+            }
+
+            // Determine if stage teardown build should start
+            // (if stage teardown build exists, and stageBuild.status is negative,
+            // and there are no active stage builds, and teardown build is not started)
+            if (stage) {
```
Comment (sagar1312) [inline plugins/builds/update.js:326]: We can skip this block, if the `newBuild` status is not a terminal status
```suggestion
            if (stage && FINISHED_STATUSES.includes(newBuild.status)) {
```

File: plugins/builds/update.js [hunk 18]
```diff
@@ -245,16 +270,99 @@ module.exports = () => ({
...
+                    if (
+                        stageTeardownBuild &&
+                        stageTeardownBuild.status === 'CREATED' &&
+                        stageBuildHasFailure &&
+                        stageIsDone
+                    ) {
+                        stageTeardownBuild.status = 'QUEUED';
+                        await stageTeardownBuild.update();
+                        await stageTeardownBuild.start();
+                    }
```
Comment (sagar1312) [inline plugins/builds/update.js:364]: If the `teardown` build is already started, we don't need to query for other stage job builds and determine if the stage is done.

Also move the logic to determine whether the stage is done to. a separate function.
```suggestion
if(stageTeardownBuild.status === 'CREATED' && isStageDone(stage)) {
    stageTeardownBuild.status = 'QUEUED';
    await stageTeardownBuild.update();
    await stageTeardownBuild.start();
}
```


```
isStageDone(stage){
// Get all jobIds for jobs in the stage
                    const stageJobIds = stage.jobIds;

                    stageJobIds.push(stage.setup);

                    // Get all builds in a stage
                    const stageJobBuilds = await buildFactory.list({
                        params: { jobId: stageJobIds, eventId: newEvent.id }
                    });
                    let stageIsDone = false;

                    if (stageJobBuilds && stageJobBuilds.length !== 0) {
                        stageIsDone = !stageJobBuilds.some(b => !FINISHED_STATUSES.includes(b.status));
                    }

return stageIsDone;
}
```

**Review by sagar1312 — CHANGES_REQUESTED** [review]

File: plugins/builds/index.js (hunk 6 above)
Comment (tkyi) [reply plugins/builds/index.js:912]: https://github.com/screwdriver-cd/models/pull/606

File: plugins/builds/index.js [hunk 19]
```diff
@@ -709,6 +721,82 @@ function getParentBuildIds({ currentBuildId, parentBuilds, joinListNames, pipeli
...
+/**
+ * Create stage teardown build if it doesn't already exist
+ * @param  {Factory}    jobFactory                      Job factory
+ * @param  {Factory}    buildFactory                    Build factory
+ * @param  {Object}     current                         Current object
+ * @param  {String}     stageTeardownName               Stage teardown name
+ * @param  {String}     username                        Username
+ * @param  {String}     scmContext                      SCM context
+ */
+async function checkStageTeardownBuild({ jobFactory, buildFactory, current, stageTeardownName, username, scmContext }) {
```
Comment (sagar1312) [inline plugins/builds/index.js:733]:
```suggestion
async function checkAndCreateStageTeardownBuild({ jobFactory, buildFactory, current, stageTeardownName, username, scmContext }) {
```

**Review by sagar1312 — APPROVED** [review]

**Review by sagar1312 — APPROVED** [review]

**Review by sagar1312 — APPROVED** [review]
